io/disk/lvsetup.py

```diff
@@ -0,0 +1,120 @@
+
+"""
+Test that automatically takes shapshots from created logical volumes
+using a given policy.
+
+For details about the policy see README.
+"""
+import avocado
+from avocado import Test
+from avocado import main
+from avocado.utils import process
+from avocado.utils import lv_utils
+
+
+class Lvsetup(Test):
+
+    """
+    Test class for creating logical volumes.
+    """
+
+    def setUp(self):
+        """
+        Check existence of input PV,VG, LV and snapshots prior to Test.
+        """
+        self.disk = self.params.get('disk', default=None)
+        vg_name = self.params.get('vg_name', default='avocado_vg')
+        self.loop = True
+        if self.disk:
+            vg_name = vg_name + '_' + self.disk.split('/')[-1]
+            self.loop = False
+        lv_name = self.params.get('lv_name', default='avocado_lv')
+        self.lv_size = self.params.get('lv_size', default='1G')
+        lv_snapshot_name = self.params.get(
+            'lv_snapshot_name', default='avocado_sn')
+        self.lv_snapshot_size = self.params.get(
+            'lv_snapshot_size', default='1G')
+        self.ramdisk_vg_size = self.params.get(
+            'ramdisk_vg_size', default='10000')
+        self.ramdisk_basedir = self.params.get(
+            'ramdisk_basedir', default=self.workdir)
+        self.ramdisk_sparse_filename = self.params.get(
+            'ramdisk_sparse_filename', default='virtual_hdd')
+
+        if lv_utils.vg_check(vg_name):
+            self.skip('Volume group %s already exists' % vg_name)
+        self.vg_name = vg_name
+        if lv_utils.lv_check(vg_name, lv_name):
+            self.skip('Logical Volume %s already exists' % lv_name)
+        self.lv_name = lv_name
+        if lv_utils.lv_check(vg_name, lv_snapshot_name):
+            self.skip('Snapshot %s already exists' % lv_snapshot_name)
+        self.lv_snapshot_name = lv_snapshot_name
+
+    @avocado.fail_on(lv_utils.LVException)
+    def test(self):
+        """
+        General logical volume setup.
+
+        A volume group with given name is created in the ramdisk. It then
+        creates a logical volume. Takes a snapshot from the logical and
+        merges snapshot with the logical volume.
+        """
+
+        self.loop_device = lv_utils.vg_ramdisk(self.disk, self.vg_name,
+                                               self.ramdisk_vg_size,
+                                               self.ramdisk_basedir,
+                                               self.ramdisk_sparse_filename)[3]
+        lv_utils.lv_create(self.vg_name, self.lv_name, self.lv_size)
+        lv_utils.lv_take_snapshot(self.vg_name, self.lv_name,
+                                  self.lv_snapshot_name,
+                                  self.lv_snapshot_size)
+        lv_utils.lv_revert(self.vg_name, self.lv_name, self.lv_snapshot_name)
+
+    def tearDown(self):
+        """
+        Clear all PV,VG, LV and snapshots created by the test.
+        """
+        # Remove created VG
```

Comment:
Hello @harish-24 this is much better, thanks. Still the cleanup is not complete and it could be actually way simpler, you could simply use `vg_ramdisk_cleanup` function, which tries to gracefully cleanup the system and raises the exception after this is done.

If, for some reason, you don't want to use this function, please add `umount` for loop device and `self.error` at the end of the cleanup if any of cleanups failed as we don't want to affect following tests by failed cleanup without a notice (and message in log is not sufficient, the test needs to fail and avocado actually notifies you that it was the cleanup what failed).